modules: nanopb: introduce new module

pdgendt opened the pull request:
Add the nanopb library and generator tools as a module.
Nanopb is a small code-size Protocol Buffers implementation in ansi C.
It is especially suitable for use in microcontrollers,
but fits any memory restricted system.

Nanopb home: https://jpa.kapsi.fi/nanopb/
License: zlib

pdgendt committed:
modules: nanopb: introduce new module

Add the nanopb library and generator tools as a module.
Nanopb is a small code-size Protocol Buffers implementation in ansi C.
It is especially suitable for use in microcontrollers,
but fits any memory restricted system.

Nanopb home: https://jpa.kapsi.fi/nanopb/

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt commented:
For the added sample to build, `protoc` and python package for Google's `protobuf` need to be installed. How should I proceed with this PR?

carlescufi commented:
> For the added sample to build, `protoc` and python package for Google's `protobuf` need to be installed. How should I proceed with this PR?

We need to update the Docker image to include the additional dependencies:
https://github.com/zephyrproject-rtos/ci-dockerfiles/tree/master/shippable

or use a temp hack like this one:
https://github.com/zephyrproject-rtos/zephyr/commit/4ea11f80e9242269f119d26b94cc76e5df858dd1

carlescufi commented on samples/modules/nanopb/src/simple.proto:
```diff
@@ -0,0 +1,8 @@
+
+message SimpleMessage {
+    required int32 lucky_number = 1;
+}
```
missing EOF

carlescufi commented:
@tejlmand the CMake files are currently in the module. I suggest we move it in-tree?

tejlmand commented:
> @tejlmand the CMake file is currently in the module. I suggest we move it in-tree?

That would indeed be very nice.
Also I have already spotted some flaws in https://github.com/zephyrproject-rtos/nanopb/blob/zephyr/zephyr/CMakeLists.txt in commit https://github.com/zephyrproject-rtos/nanopb/commit/1aa945e1523cb2c705ab608e7698e6d7ab4e1ac2 that I would like to get fixed before merging this PR.

tejlmand reviewed (changes requested):
This seems to require a little more work.

I did run `python -mpip install -r scripts/requirements.txt` to ensure I got the updated `protobuf` package in this PR.

But I get:
```
-- Found Nanopb: /projects/github/ncs/modules/lib/nanopb
Failed to build nanopb_pb2.py: protoc --python_out=/projects/github/ncs/modules/lib/nanopb/generator/proto /projects/github/ncs/modules/lib/nanopb/generator/proto/nanopb.proto -I=/projects/github/ncs/modules/lib/nanopb/generator/proto -I/projects/github/ncs/modules/lib/nanopb/generator/proto
Failed to build nanopb_pb2.py: protoc --python_out=/projects/github/ncs/modules/lib/nanopb/generator/proto /projects/github/ncs/modules/lib/nanopb/generator/proto/nanopb.proto -I=/projects/github/ncs/modules/lib/nanopb/generator/proto -I/projects/github/ncs/modules/lib/nanopb/generator/proto
Traceback (most recent call last):
  File "/projects/github/ncs/modules/lib/nanopb/generator/nanopb_generator.py", line 51, in <module>
    from .proto import nanopb_pb2
ImportError: attempted relative import with no known parent package

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/projects/github/ncs/modules/lib/nanopb/generator/protoc", line 6, in <module>
    from nanopb_generator import invoke_protoc
  File "/projects/github/ncs/modules/lib/nanopb/generator/nanopb_generator.py", line 72, in <module>
    import proto.nanopb_pb2 as nanopb_pb2
  File "/projects/github/ncs/modules/lib/nanopb/generator/proto/__init__.py", line 33, in <module>
    invoke_protoc(argv=cmd)
  File "/projects/github/ncs/modules/lib/nanopb/generator/proto/_utils.py", line 43, in invoke_protoc
    return subprocess.call(argv)
  File "/usr/lib/python3.8/subprocess.py", line 340, in call
    with Popen(*popenargs, **kwargs) as p:
  File "/usr/lib/python3.8/subprocess.py", line 854, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib/python3.8/subprocess.py", line 1702, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'protoc'
CMake Error at /projects/github/ncs/modules/lib/nanopb/extra/FindNanopb.cmake:263 (string):
  string sub-command REGEX, mode MATCH needs at least 5 arguments total to
  command.
Call Stack (most recent call first):
  CMakeLists.txt:8 (nanopb_generate_cpp)

```
Especially `string sub-command REGEX, mode MATCH needs at least 5 arguments total to
  command.` will be cryptic to most users.

Downloading https://jpa.kapsi.fi/nanopb/download/nanopb-0.4.5-linux-x86.tar.gz, extracting and adding `<path>/nanopb-0.4.5-linux-x86/generator-bin/` to my system path helps a lot, and i'm able to build the sample succesfully.
But there is no documentation informing users of this step.

This current docs has no mentioning of this: https://github.com/zephyrproject-rtos/zephyr/blob/f84e047e77c293c0da4aa6bf1a847616a108c327/samples/modules/nanopb/README.rst

So before approving, I would like to see:
- Better documentation, describing what the user needs to do in order to build: `samples/modules/nanopb`
- A better error message if user forgot to do the required step.
  Having first a: `-- Found Nanopb: /projects/github/ncs/modules/lib/nanopb` followed by a python stack trace ending with: `FileNotFoundError: [Errno 2] No such file or directory: 'protoc'` and then topped with: `string sub-command REGEX, mode MATCH needs at least 5 arguments total to command.` is simply not good enough for people, and will result in too many users waisting too much time.

pdgendt commented:
Thanks @tejlmand for the review!

I moved the cmake in-tree, so if you have comments on that part, I'd be happy to improve it!

As for the cryptic string error, that was a direct consequence of the previous error, being that the `protoc` executable could not be found.
However, this is all output from the cmake file that nanopb itself maintains, so I'm not sure how to change this. We could move the cmake functions to the zephyr tree, but I probably lack the cmake skills to do it properly, so please advise.

I also added the `protoc` requirement to the sample README, is that sufficient?

carlescufi commented:
> As for the cryptic string error, that was a direct consequence of the previous error, being that the `protoc` executable could not be found.

It's not *that* cryptic:
`FileNotFoundError: [Errno 2] No such file or directory: 'protoc'`

> However, this is all output from the cmake file that nanopb itself maintains, so I'm not sure how to change this. We could move the cmake functions to the zephyr tree, but I probably lack the cmake skills to do it properly, so please advise.

I don't think we should modify nanopb's CMake files unless we send those changes upstream, because we are talking about changes that are not particular to Zephyr. But this is not a CMake problem is it?

`ImportError: attempted relative import with no known parent package`

pdgendt commented:
> It's not _that_ cryptic:
> `FileNotFoundError: [Errno 2] No such file or directory: 'protoc'`

No, but the `string sub-command REGEX, mode MATCH needs at least 5 arguments total to command` as @tejlmand pointed out, is. But I agree that the preceding error is clear.

> I don't think we should modify nanopb's CMake files unless we send those changes upstream, because we are talking about changes that are not particular to Zephyr. But this is not a CMake problem is it? It looks like a Python one:
> 
> `ImportError: attempted relative import with no known parent package`

Looking at the generator code, it verifies some fallback scenarios and runs those in the `except` clause. That is why the import error is printed, but I think it works as intended.

Are there still issues if the requirements (`protoc` and python `protobuf`) are met? Also how can I fix the buildkite build?

tejlmand commented:
> > It's not _that_ cryptic:
> > `FileNotFoundError: [Errno 2] No such file or directory: 'protoc'`
> 
> No, but the `string sub-command REGEX, mode MATCH needs at least 5 arguments total to command` as @tejlmand pointed out, is. But I agree that the preceding error is clear.

Yes, for users like us who knows how to relate those messages together, it's clear.
But I wrote: 
> Especially string sub-command REGEX, mode MATCH needs at least 5 arguments total to command. will be cryptic to most users.
and most users may not know that the CMake error is an indirect cause of the python error.

Also, a user that does:
```
west clone
west update
pip install -s scripts/requirements.txt
west build  samples/modules/nanopb
```

how would he be aware that he must install additional tools ?
This is not mentioned anywhere in the docs.

> > However, this is all output from the cmake file that nanopb itself maintains, so I'm not sure how to change this. We could move the cmake functions to the zephyr tree, but I probably lack the cmake skills to do it properly, so please advise.
> 
> I don't think we should modify nanopb's CMake files unless we send those changes upstream, because we are talking about changes that are not particular to Zephyr. But this is not a CMake problem is it? It looks like a Python one:
> 
> `ImportError: attempted relative import with no known parent package`

I'm not talking about changing upstream, although I believe it would be a nice gesture to send a PR or raise an issue there.
What should be done in the Zephyr nanopb CMake code is to have a `find_program(PROTOC protoc)` and if that  fails, then provide proper explanation to the user what is wrong, and if possible, how to fix it.

pdgendt commented:
> I'm not talking about changing upstream, although I believe it would be a nice gesture to send a PR or raise an issue there.
> What should be done in the Zephyr nanopb CMake code is to have a `find_program(PROTOC protoc)` and if that fails, then provide proper explanation to the user what is wrong, and if possible, how to fix it.

Okay, further investigation shows that using the python package `grpcio-tools` does provide the protocol buffer compiler, which eliminates the `protoc` executable requirement.
I will add this to `requirements-extras.txt`, which is probably the easiest/cleanest solution?

carlescufi commented:
> Okay, further investigation shows that using the python package `grpcio-tools` does provide the protocol buffer compiler, which eliminates the `protoc` executable requirement.
> I will add this to `requirements-extras.txt`, which is probably the easiest/cleanest solution?

Well, it's either or. Both `protoc` and `grpcio-tools` provide a protobuffers compiler. Since both options are supported upstream in nanopb, I think we should choose one as the standard one we document and go with for our users. If we decide to go with `grpcio-tools`, then we need to update the CI Docker image with it, otherwise we'd be testing with `protoc` but asking our users to use `grpcio-tools`. But given that this can be solved in Python only, let's go with that. Add the requirement and this should then do away with the need for `protoc` altogether and should make it nicer for the user.

carlescufi commented:
Actually @pdgendt, to me it looks like `grpcio-tools` also requires `protoc` to be installed:
https://github.com/grpc/grpc/tree/master/tools/distrib/python/grpcio_tools

Could you please confirm this, and document it if it's the case?

carlescufi commented:
> Actually @pdgendt, to me it looks like `grpcio-tools` also requires `protoc` to be installed:
> https://github.com/grpc/grpc/tree/master/tools/distrib/python/grpcio_tools
> 
> Could you please confirm this, and document it if it's the case?

Actually, I might be wrong. It might actually pull the `protoc` code directly into the python package.

carlescufi reviewed (changes requested):
There is currently an issue with using `grpcio-tools` on Windows, which needs to be solved upstream:
```
CMake Error at /projects/github/ncs/modules/lib/nanopb/extra/FindNanopb.cmake:263 (string):
  string sub-command REGEX, mode MATCH needs at least 5 arguments total to
```

In order to save users from confusion, document explicitly that one must use `protoc`, at least on Windows. On Linux, both `protoc` and `grpcio-tools` seem to work.

Currently the Windows build is also broken, @tejlmand can take a look at this:

```
-- west build: building application
[98/147] Building C object CMakeFiles/app.dir/src/simple.pb.c.obj
FAILED: CMakeFiles/app.dir/src/simple.pb.c.obj
C:\gnuarmemb\bin\arm-none-eabi-gcc.exe -DBUILD_VERSION=zephyr-v2.5.0-1640-gd15b12fc50f3 -DKERNEL -DNRF52840_XXAA -DPB_MAX_REQUIRED_FIELDS=64 -D_FORTIFY_SOURCE=2 -D__PROGRAM_START -D__ZEPHYR__=1 -IC:/Users/Carles/src/zephyr/zephyr/include -Izephyr/include/generated -IC:/Users/Carles/src/zephyr/zephyr/soc/arm/nordic_nrf/nrf52 -IC:/Users/Carles/src/zephyr/modules/hal/cmsis/CMSIS/Core/Include -IC:/Users/Carles/src/zephyr/modules/hal/nordic/nrfx -IC:/Users/Carles/src/zephyr/modules/hal/nordic/nrfx/drivers/include -IC:/Users/Carles/src/zephyr/modules/hal/nordic/nrfx/mdk -IC:/Users/Carles/src/zephyr/zephyr/modules/hal_nordic/nrfx/. -IC:/Users/Carles/src/zephyr/modules/debug/segger/rtt -IC:/Users/Carles/src/zephyr/modules/lib/nanopb/spm_headers/nanopb -I. -isystem C:/Users/Carles/src/zephyr/zephyr/lib/libc/minimal/include -isystem c:/gnuarmemb/bin/../lib/gcc/arm-none-eabi/9.3.1/include -isystem c:/gnuarmemb/bin/../lib/gcc/arm-none-eabi/9.3.1/include-fixed -Os -imacros C:/Users/Carles/src/zephyr/zephyr/build/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -mcpu=cortex-m4 -mthumb -mabi=aapcs -imacros C:/Users/Carles/src/zephyr/zephyr/include/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-address-of-packed-member -Wno-unused-but-set-variable -Werror=implicit-int -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=C:/Users/Carles/src/zephyr/zephyr/samples/modules/nanopb=CMAKE_SOURCE_DIR -fmacro-prefix-map=C:/Users/Carles/src/zephyr/zephyr=ZEPHYR_BASE -fmacro-prefix-map=C:/Users/Carles/src/zephyr=WEST_TOPDIR -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT CMakeFiles/app.dir/src/simple.pb.c.obj -MF CMakeFiles\app.dir\src\simple.pb.c.obj.d -o CMakeFiles/app.dir/src/simple.pb.c.obj   -c src/simple.pb.c
In file included from ./src/simple.pb.h:6,
                 from src/simple.pb.c:4:
C:/Users/Carles/src/zephyr/modules/lib/nanopb/spm_headers/nanopb/pb.h:1:1: error: expected identifier or '(' before '.' token
    1 | ../../pb.h
      | ^
In file included from src/simple.pb.c:4:
./src/simple.pb.h:9:2: error: #error Regenerate this file with the current version of nanopb generator.
    9 | #error Regenerate this file with the current version of nanopb generator.
      |  ^~~~~
```
It's broken because there are symlinks here:
`modules\lib\nanopb\spm_headers\nanopb`
and symlinks do not work on Windows. Git insead creates a file with the symlink destination as its contents.

See this comment for more info:
https://github.com/zephyrproject-rtos/zephyr/pull/33211/files#r603401287

carlescufi commented on modules/nanopb/CMakeLists.txt:
```diff
@@ -0,0 +1,54 @@
+
+  zephyr_library()
+
+  zephyr_include_directories(${NANOPB_DIR}/spm_headers/nanopb)
```
This will not work on Windows, because it uses symlinks. Avoid the `spm_headers` folder completely and instead point to `modules/lib/nanopb`.

pdgendt replied:
Yes, reading your previous comment, I figured that would be the issue, will edit.

tejlmand commented on scripts/requirements-extras.txt:
```diff
@@ -17,3 +17,7 @@ Pillow
+
+# used by nanopb module to generate sources from .proto files
+protobuf
+grpcio-tools
```
Unfortunately this turned out to not be sufficient for Windows users, so there is still need to ensure that `protoc` is in path to make it easier for user to understand the error.

tejlmand commented on modules/nanopb/CMakeLists.txt:
```diff
@@ -0,0 +1,54 @@
+
+  set(NANOPB_DIR ${ZEPHYR_CURRENT_MODULE_DIR})
+
+  include(${NANOPB_DIR}/extra/FindNanopb.cmake)
```
This gives a warning for CMake 3.17 and newer.

Please use:
```suggestion
  set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} ${NANOPB_DIR}/extra)
  find_package(Nanopb REQUIRED)
```

Note, warning seen:
```
CMake Warning (dev) at /opt/cmake-3.19.1-Linux-x86_64/share/cmake-3.19/Modules/FindPackageHandleStandardArgs.cmake:424 (message):
  The package name passed to `find_package_handle_standard_args` (Nanopb)
  does not match the name of the calling package (Zephyr).  This can lead to
  problems in calling code that expects `find_package` result variables
  (e.g., `_FOUND`) to follow a certain pattern.
Call Stack (most recent call first):
  /projects/github/ncs/modules/lib/nanopb/extra/FindNanopb.cmake:365 (find_package_handle_standard_args)
  ../../../modules/nanopb/CMakeLists.txt:9 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.
```

pdgendt commented on scripts/requirements-extras.txt:
```diff
@@ -17,3 +17,7 @@ Pillow
+
+# used by nanopb module to generate sources from .proto files
+protobuf
+grpcio-tools
```
So the `protoc` executable is the preferred way to go probably, I've modified the sample `README` on how to install it on the different platforms.
Is this sufficient, or do we need to do a check as well?

tejlmand replied:
See suggestion here: https://github.com/zephyrproject-rtos/zephyr/pull/33211#discussion_r604639622

tejlmand reviewed (changes requested):
still would like an extra check. `protoc` is not among the default tools or in requirements-*.txt file, combined with the not so clear message from Nanopb.

I believe this can save time for users.

jfischer-no commented on samples/modules/nanopb/src/simple.proto:
```diff
@@ -0,0 +1,8 @@
+
+syntax = "proto2";
+
+message SimpleMessage {
```
Can you use Zephyr's coding style here?

jfischer-no replied:
You mean if one change it to `simple_message {` it would stop to work? Should it then be in Zephyr's tree?

pdgendt replied:
It wouldn't stop working, it only affects the generated source/header files, and for languages other than `C/C++` it would produce potentially weird syntax.

This only is a sample, so I wanted to keep it as close to the samples provided by google/nanopb, because I think it would be more common to look at those as a reference.

tejlmand reviewed (approved):
lgtm

carlescufi commented on samples/modules/nanopb/src/main.c:
```diff
@@ -0,0 +1,77 @@
+#include <pb_decode.h>
+#include "src/simple.pb.h"
+
+int main(void)
```
@pdgendt `void main(void)`

carlescufi commented on samples/modules/nanopb/src/main.c:
```diff
@@ -0,0 +1,77 @@
+	bool status;
+
+	/* Encode our message */
+	{
```
Can you please remove these scope brackets, and the ones below? This is not something we usually make use of.

pdgendt replied:
I've moved the scoped logic to functions, this is to demonstrate the stack usage of the message struct.

mniestroj commented on modules/nanopb/Kconfig:
```diff
@@ -0,0 +1,59 @@
+	default 64
+	help
+	  Maximum number of proto2 required fields to check for presence.
+	  Default value is 64. Compiler warning will tell if you need this.
```
... if you need **to change** this. ?
or
... if you need **to increase** this. ?

pdgendt replied:
I used the same documentation as the original repo, but it is actually just the minimum value, and the compiler will complain if it is set lower than 64.

I will update the description.

pdgendt commented:
@carlescufi I added the test section for the sample, but this will fail on systems that do not have the `protoc` executable installed. Can this be checked somehow to skip the test in that case?

carlescufi commented:
> @carlescufi I added the test section for the sample, but this will fail on systems that do not have the `protoc` executable installed. Can this be checked somehow to skip the test in that case?

No, the assumption is that you run this with the Zephyr SDK, which contains it. So nothing else to do here.
